rfcs/0008-transition-frontier-controller.md

```diff
@@ -0,0 +1,337 @@
+we've never explored in the existing ledger-builder-controller (what happens in
+the presense of particular looking forks).
+
+One alternative is to instead prioritize in investing in debugging
+infrastructure such as visualizations and logging tooling to get through these
+asynchronous bugs. However, we will need to do a medium-sized refactor anyway to integrate the
+merkle masks properly and the new correct history catchup anyway, so we're not
+actually saving much work.
+
+This seems to be a good point to do a full-rewrite given that we have these new
+components we wish to hook in and we also are running into debugging issues and
+new bugs.
+
+## Prior art
+[prior-art]: #prior-art
+
+The main piece of prior art is the existing Ledger-builder-controller
+component. Ledger-builder-controller as is supports the same external interface
+and a lot of similar behavior, but in a more adhoc way. The existing
+ledger-builder-controller handles incoming transitions by spawning one of two
+types of asynchronous jobs, a catchup and a path traversal. As a short-cut,
+catchup did not differentiate between the "from nothing" case and the "small
+miss" case, and was vulnerable to an attack. Path traversal was the process of
+materializing a ledger-builder along a path in the ktree. This had to be
+asynchronous because we didn't yet have a notion of a masked ledger. The
+ledger-builder-controller kept only one async job running at a time for
+simplicity as well, and only cancelled in-progress jobs when another should
+replace the existing one. Upon the completion of a job, the
+ledger-builder-controller would replace it's ktree and update annotated tip
+entities that have materialized ledger-builders in them.
+Ledger-builder-controller also handled sync-ledger answers as it had to access
+the path traversal logic and the current transition logic state.
+
+This component grew organically and did not have any rigorous design
+thought put behind it as we did not realize how central such a component would
+be in our system until it was too late.
+
+Moreover, we did not think carefully about the flow of information through this
+component and the implications of various async points from getting backed up.
+
+The bits we'll keep: We have separate and well-unit-tested component for the
+`ktree` and have some decent integration-esque tests around the
+ledger-builder-controller as a whole that we should be able to mostly reuse.
+We're also reusing the only-one-at-a-time shortcut for our async jobs for now.
+
+## Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+### High-throughput
+
+If we want to support throughput higher than ~10tps we'll need to change
+breadcrumb storage to be backed by some sort of disk-backed store.
+
+This is not expected to be resolved before landing this RFC.
+
+### History catchup
+
+@es92 notes that there is some nuance to catching up to breadcrumbs for
```

Comment:
@es92 I would like to document some of this nuance to understand it better, especially in the context of this RFC.

Reply:
Basically we should plan and document a system that
1. has a way to download diffs back to the locked tip from a disconnected state (who will that be downloaded from?)
2. can walk back the chain for multiple disconnected chains in parallel
3. can accept children of disconnected chains while the chains are being connected up

Reply:
oh great we've discussed these all already